api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,818 @@
+		switch s.SortBy {
+		case "cpu":
+			sortProvidersByCPU(s.Providers)
+			sortConsumersByCPU(resourceConsumers)
+
+		case "mem":
+			sortProvidersByMemory(s.Providers)
+			sortConsumersByMemory(resourceConsumers)
+		}
+
+		sortProvidersByUsage(s.Providers)
+
+		for _, consumer := range resourceConsumers {
+			hasRoom := false
+
+			for _, provider := range s.Providers {
+				if provider.HasResourcesFor(consumer) {
+					hasRoom = true
+					if err := provider.SubtractResourcesFor(consumer); err != nil {
+						s.Errs = append(s.Errs, err)
+					}
+
+					break
+				}
+			}
+
+			if !hasRoom {
+				newProvider, err := e.calculateNewProvider(clusterName)
+				if err != nil {
+					return nil, nil, err
+				}
+
+				if !newProvider.HasResourcesFor(consumer) {
+					text := fmt.Sprintf("Resource '%s' cannot fit into an empty provider!", consumer.ID)
+					text += "\nThe instance size in your environment is too small to run this resource."
+					text += "\n Please increase the instance size for your environment."
+					err := fmt.Errorf(text)
+					s.Errs = append(s.Errs, err)
+					continue
+				}
+
+				if err := newProvider.SubtractResourcesFor(consumer); err != nil {
+					s.Errs = append(s.Errs, err)
+				}
+
+				s.Providers = append(s.Providers, newProvider)
+			}
+		}
+	}
+
+	p := findOptimalProviderDistribution(sorts)
+
+	return p.Providers, p.Errs, nil
+}
+
+func findOptimalProviderDistribution(providerDistributions []ProviderDistribution) ProviderDistribution {
+	if len(providerDistributions) == 1 {
+		return providerDistributions[0]
+	}
+
+	shortestLength := len(providerDistributions[0].Providers)
+	shortestDistribution := providerDistributions[0]
+
+	for _, p := range providerDistributions[1:] {
+		l := len(p.Providers)
+		if l < shortestLength {
+			shortestLength = l
+			shortestDistribution = p
+		}
+	}
+
+	return shortestDistribution
+}
+
+func (e *EnvironmentScaler) getContainerResourceFromDeploy(deployID string) ([]scaler.ResourceConsumer, error) {
+	if consumers, ok := e.deployCache.Getf(deployID); ok {
+		return consumers.([]scaler.ResourceConsumer), nil
+	}
+
+	input := &ecs.DescribeTaskDefinitionInput{}
+	input.SetTaskDefinition(deployID)
+	output, err := e.Client.ECS.DescribeTaskDefinition(input)
+	if err != nil {
+		return nil, err
+	}
+
+	consumers := make([]scaler.ResourceConsumer, len(output.TaskDefinition.ContainerDefinitions))
+	for i, d := range output.TaskDefinition.ContainerDefinitions {
+		var memory bytesize.Bytesize
+
+		if aws.Int64Value(d.MemoryReservation) != 0 {
+			memory = bytesize.MiB * bytesize.Bytesize(aws.Int64Value(d.MemoryReservation))
+		}
+
+		if aws.Int64Value(d.Memory) != 0 {
+			memory = bytesize.MiB * bytesize.Bytesize(aws.Int64Value(d.Memory))
+		}
+
+		ports := []int{}
+		for _, p := range d.PortMappings {
+			if aws.Int64Value(p.HostPort) != 0 {
+				ports = append(ports, int(aws.Int64Value(p.HostPort)))
+			}
+		}
+
+		consumers[i] = scaler.ResourceConsumer{
+			ID:     "",
```

Comment:
If this is the 0-value for `ID` then I don't think we need to specify it here.

Reply:
There's an attempt to provide this a better value for debugging purposes as per another suggestion, but otherwise I agree. ✅